src/dodal/plans/get_run_engine_metadata.py

```diff
@@ -0,0 +1,62 @@
+from typing import Generic, Protocol, TypeVar
+
+import bluesky.plan_stubs as bps
+import bluesky.preprocessors as bpp
+from bluesky.callbacks import CallbackBase
+from bluesky.utils import MsgGenerator
+
+
+class FromStr(Protocol):
+    def __init__(self, value: str) -> None: ...
+
+
+T = TypeVar("T", str, int, float)
+
+
+class _GetRunEngineMetaDataCallback(CallbackBase, Generic[T]):
+    def __init__(self, metadata_name: str, expected_datatype: type[T] = str):
+        self.metadata: T | None = None
+        self.expected_datatype = expected_datatype
+        self.metadata_name = metadata_name
+        super().__init__()
+
+    def start(self, doc):
+        self.metadata = doc.get(self.metadata_name)
+        if not self.metadata:
+            raise ValueError(
+                f"Requested RunEngine metadata '{self.metadata_name}' could not be found"
+            )
+        try:
+            self.metadata = self.expected_datatype(self.metadata)
+        except Exception as e:
+            raise TypeError(
+                f"Requested RunEngine metadata '{self.metadata}' could not be converted to requested type '{self.expected_datatype}'"
+            ) from e
+
+
+def get_run_engine_metadata(
+    metadata_name: str, expected_datatype: type[T] = str
+) -> MsgGenerator[T]:
+    """
+    Get metadata stored on the RunEngine.
+
+    Creates a callback and a dummy run event to read metadata from the RunEngine. Useful
+    for plans which don't have reference to the RunEngine, for example when running through
+    BlueAPI
+
+    Args:
+    metadata_name: The string name for requested metadata
+    expected_datatype: The datatype you expect the data to be in. String by default.
+    """
+
+    callback = _GetRunEngineMetaDataCallback[T](metadata_name, expected_datatype)
+
+    @bpp.subs_decorator(callback)
+    @bpp.run_decorator()
```

Comment:
Won't making this a run in its own right end up with it getting its own scan numbers/run IDs etc with a call to numtracker for each call? Does creating nested runs cause issues in any of our downstream handling (tiled/nexus writer etc)?

If exposing run_engine metadata is something that is not possible with the run engine builtins, would it be worth registering a new `metadata` message handler so you could do something like

```
instrument = yield Msg("metadata", key="instrument")
```
Is there anything in the metadata that we don't want plans to be able to access?